src/hotspot/share/gc/shenandoah/heuristics/shenandoahOldHeuristics.cpp

```diff
@@ -202,6 +202,19 @@ void ShenandoahOldHeuristics::prepare_for_old_collections() {
         size_t region_count = heap->trash_humongous_region_at(region);
         log_debug(gc)("Trashed " SIZE_FORMAT " regions for humongous object.", region_count);
       }
+    } else if (region->is_pinned()) {
+      assert(!region->is_humongous(), "Humongous region should be handled elsewhere.");
+      if (region->has_live()) {
+        // This region is pinned, so we aren't going to include it in the collection set.
+        // However, we must still 'fill' in the dead objects in this region to stop
+        // subsequent remembered set scan from tracing through into garbage. Here we add it
+        // to the set of candidates, but with _zero_ garbage so that it sorts to the end of
+        // the garbage-first list.
+        region->begin_preemptible_coalesce_and_fill();
+        candidates[cand_idx]._region = region;
+        candidates[cand_idx]._garbage = 0;
+        cand_idx++;
+      }
```

Comment:
If it is selected as a candidate for an old evacuation, access to it will be protected by the mark bitmap (`heap->doing_mixed_evacuations()` will be true).

Hmm, we might not handle that case properly. Independent of the filling dead objects issue, I can see nothing to prevent old regions with a `pin_count() > 0` from being added to a mixed collection (which would be bad for other reasons). I shall convert this PR to draft until we have a solution that handles that scenario. Perhaps we can also address Ramki's concern that regions that become 'unpinned' before the mixed collection _should_ be included.

Reply:
The root of the issue is that mutator threads change `pin_count`, but only the gc threads transition (or synchronize) the pin count with `pinned` or `unpinned` states.

Reply:
Modulo yr concerns re the transitions into and out of `_pinned` state, I'd have thought that `_cset_pinned` could be considered the moral equivalent of not being in `_cset`. I'd expect that if an attempt were made to pin an object in a region that is in state `_cset`, the caller (mutator) would be forced to evacuate the object and then pin the destination region before the pinning call returns?

Reply:
I would hope there is some existing code that rejects a region-pin request if the region has already been selected to be in the "cset".  Actually, this probably will never happen because once a region has been selected to be in cset, then the mutator can no longer see the objects residing within that region.  The mutator can only see the to-space replicas of these objects.  So the mutator may try to pin the destination region, but will never try to pin a from-space (cset) region.

Reply:
@kdnilsen - You're right that the mutator _should_ not be allowed to pin regions in the cset. The load barrier should evacuate the object and the to-space region would be pinned. However, if there is an OOM failure during the evacuation, the mutator will be able to reach the from-space object and pin its region. This scenario is [handled in the degenerated cycle](https://github.com/openjdk/shenandoah/blob/ca939b69f95db2cb8ead9c9459b2d7712909a4b6/src/hotspot/share/gc/shenandoah/shenandoahDegeneratedGC.cpp#L171):

```Java
        // Before doing that, we need to make sure we never had any cset-pinned
        // regions. This may happen if allocation failure happened when evacuating
        // the about-to-be-pinned object, oom-evac protocol left the object in
        // the collection set, and then the pin reached the cset region. If we continue
        // the cycle here, we would trash the cset and alive objects in it. To avoid
        // it, we fail degeneration right away and slide into Full GC to recover.
```

Reply:
Good catch.  So we cannot entirely avoid this issue.